### PR TITLE
Fix `MapArrayReader` (#2484) (#1699) (#1561)

### LINKED ISSUE #1561: Improve Unit Test Coverage of MapArrayReader
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

There is no unit test coverage of `MapArrayReader` that I can find

**Describe the solution you'd like**

We should have some unit tests

### LINKED ISSUE #1699: MapArrayReader Does Not Understand Nesting
**Describe the bug**

The logic within MapArrayReader will not work if the MapArray is itself nested, in particular:

* It does not expose the def_level / rep_levels of its children
* It assumes that a repetition level of 0 corresponds to a new map entry

**To Reproduce**

A nullable StructArray with a child MapArray will likely result in an error, similarly a ListArray with a child MapArray will likely error

**Expected behavior**

This should be properly supported

### LINKED ISSUE #2484: MapArray columns don't handle null values correctly
**Describe the bug**

Given a parquet file with the following nested map content:

```
 +--------------------------------------------------------+
 |map                                                     |
 +--------------------------------------------------------+
 |null                                                    |
 |null                                                    |
 |{three -> 3, four -> 4, five -> 5, six -> 6, seven -> 7}|
 +--------------------------------------------------------+
```

Reading the keys values for the third record gives the wrong data. We expect `["three", "four", "five", "six", "seven"]` but get `[null, null, "three", "four", "five"]`.

**To Reproduce**

The following test will reproduce this:

```
#[test]
// This test writes a parquet file with the following data:
// +--------------------------------------------------------+
// |map                                                     |
// +--------------------------------------------------------+
// |null                                                    |
// |null                                                    |
// |{three -> 3, four -> 4, five -> 5, six -> 6, seven -> 7}|
// +--------------------------------------------------------+
//
// It then attempts to read the data back and checks that the third record
// contains the expected values.
fn read_map_array_column() {
    // Make sure generated parquet file is removed whether test passes or not
    let _file_remover = FileRemover {
        filename: "read_map_column_with_leading_nulls.parquet".to_string(),
    };

    // Schema for single map of string to int32
    let path = Path::new("read_map_column_with_leading_nulls.parquet");
    let schema = Schema::new(vec![Field::new(
        "map",
        DataType::Map(
            Box::new(Field::new(
                "entries",
                DataType::Struct(vec![
                    Field::new("keys", DataType::Utf8, false),
                    Field::new("values", DataType::Int32, true),
                ]),
                false,
            )),
            false, // Map field not sorted
        ),
        true,
    )]);

    // Create builders for map
    let string_builder = StringBuilder::new(5);
    let ints_builder: PrimitiveBuilder<Int32Type> = PrimitiveBuilder::new(1);
    let mut map_builder = MapBuilder::new(None, string_builder, ints_builder);

    // Add two null records and one record with five entries
    map_builder.append(false).expect("adding null map entry");
    map_builder.append(false).expect("adding null map entry");
    map_builder.keys().append_value("three");
    map_builder.keys().append_value("four");
    map_builder.keys().append_value("five");
    map_builder.keys().append_value("six");
    map_builder.keys().append_value("seven");

    map_builder.values().append_value(3);
    map_builder.values().append_value(4);
    map_builder.values().append_value(5);
    map_builder.values().append_value(6);
    map_builder.values().append_value(7);
    map_builder.append(true).expect("adding map entry");

    // Create record batch
    let batch =
        RecordBatch::try_new(Arc::new(schema), vec![Arc::new(map_builder.finish())])
            .expect("create record batch");

    // Write record batch to file
    let props = WriterProperties::builder().build();
    let file = fs::File::create(&path).expect("create file");
    let mut writer = ArrowWriter::try_new(file, batch.schema(), Some(props))
        .expect("creat file writer");
    writer.write(&batch).expect("writing file");
    writer.close().expect("close writer");

    // Read file
    let file = fs::File::open(&Path::new("read_map_column_with_leading_nulls.parquet"))
        .expect("open file");
    let record_batch_reader = ParquetRecordBatchReaderBuilder::try_new(file).expect("Trying to read parquet file").build().expect("Getting batch iterator");
    for maybe_record_batch in record_batch_reader {
        let record_batch = maybe_record_batch.expect("Getting current batch");
        let col = record_batch.column(0);
        let map_entry = array::as_map_array(col).value(2);
        let struct_col = array::as_struct_array(&map_entry);
        let key_col = array::as_string_array(struct_col.column(0)); // Key column
        assert_eq!(key_col.value(0), "three");
        assert_eq!(key_col.value(1), "four");
        assert_eq!(key_col.value(2), "five");
        assert_eq!(key_col.value(3), "six");
        assert_eq!(key_col.value(4), "seven");
    }
    println!("finished reading");
    fs::remove_file("read_map_column_with_leading_nulls.parquet").expect("Removing file");
}
```

**Expected behavior**

Reading the keys values for the third record should give`["three", "four", "five", "six", "seven"]`.

**Additional context**

Debugging the code shows that the offsets array for the MapArray's key column is this `[0, 0, 0, 5, 9, 13]` when I think it should be this `[0, 5, 9, 13, 16, 21]`. It looks like the issue is in `parquet::arrow::array_reader::map_array::MapArrayReaderconsume_batch` where the repetition and definition levels are used to determine the data offsets (starting at parquet\src\arrow\array_reader\map_array.rs line 123).

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #2484 
Closes #1699
Closes #1561

# Rationale for this change
 

The ListArrayReader logic is better tested, more complete and can be used to read MapArrays

# What changes are included in this PR?

Replaces the logic in MapArrayReader with ListArrayReader

# Are there any user-facing changes?

No

### CODE REVIEW

File: parquet/src/arrow/array_reader/map_array.rs [hunk 1]
```diff
@@ -65,131 +85,122 @@ impl ArrayReader for MapArrayReader {
...
+    use super::*;
+    use crate::arrow::arrow_reader::ParquetRecordBatchReader;
+    use crate::arrow::ArrowWriter;
+    use arrow::array;
+    use arrow::array::{MapBuilder, PrimitiveBuilder, StringBuilder};
+    use arrow::datatypes::{Field, Int32Type, Schema};
+    use arrow::record_batch::RecordBatch;
+    use bytes::Bytes;
+
+    #[test]
```
Comment (tustvold) [inline parquet/src/arrow/array_reader/map_array.rs:124]: Credit to @d-willis for this test

File: parquet/src/arrow/array_reader/map_array.rs [hunk 2]
```diff
@@ -43,15 +35,43 @@ impl MapArrayReader {
...
+            true => def_level + 2,
+            false => def_level + 1,
+        };
+        let struct_rep_level = rep_level + 1;
+
+        let element = match &data_type {
+            ArrowType::Map(element, _) => match element.data_type() {
+                ArrowType::Struct(fields) if fields.len() == 2 => {
+                    // The inner map field must always non-nullable (#1697)
+                    assert!(!element.is_nullable(), "map struct cannot be nullable");
```
Comment (tustvold) [inline parquet/src/arrow/array_reader/map_array.rs:50]: As described in #1697 the encoding of DataType::Map permits greater nullability than the type can actually contain

**Comment by tvincent2** [comment]
Hi!
You just made my week with this PR!

It's quite a coincidence, but we've been hit by this very problem and spent some time on it. We created [a small project](https://github.com/pbstck/arrow-issue) to show how to reproduce our issue and just found #2484 as we were about to open our issue.
We can confirm that we can no longer reproduce our issue when using this PR branch.

Thanks a lot!

Cheers,
Thomas

**Comment by d-willis** [comment]
Giving this a test against our code now. Hopefully will have run by tomorrow...
I can confirm that the reproducer test now works correctly locally.

File: parquet/src/arrow/array_reader/map_array.rs [hunk 3]
```diff
@@ -65,131 +85,122 @@ impl ArrayReader for MapArrayReader {
...
+        writer.close().expect("close writer");
+
+        // Read file
+        let reader = Bytes::from(buffer);
+        let record_batch_reader =
+            ParquetRecordBatchReader::try_new(reader, 1024).unwrap();
+        for maybe_record_batch in record_batch_reader {
+            let record_batch = maybe_record_batch.expect("Getting current batch");
+            let col = record_batch.column(0);
+            let map_entry = array::as_map_array(col).value(2);
```
Comment (alamb) [inline parquet/src/arrow/array_reader/map_array.rs:196]: Shouldn't this test also validate that the first two entries are null? Something like

```rust
assert!(!array::as_map_array(col).is_valid(0));
assert!(!array::as_map_array(col).is_valid(1));
```

**Review by alamb — APPROVED** [review]
Thank you @tustvold  - I think this looks really nice (it is awesome to fix bugs by deleting code). 

I think the test  should be updated to assert that `nullness` on the read array was  correct, but otherwise this PR looks good to go to me.

The feedback from @tvincent2 (thanks for that btw 👋 )  also makes a compelling case for the benefits of this PR. 👍 

cc @nevi-me and @viirya

File: parquet/src/arrow/array_reader/map_array.rs (hunk 3 above)
Comment (d-willis) [reply parquet/src/arrow/array_reader/map_array.rs:196]: The test was originally written to test my very specific failure condition.

Comment (d-willis) [reply parquet/src/arrow/array_reader/map_array.rs:196]: Possibly a worthwhile addition though.

File: parquet/src/arrow/array_reader/map_array.rs (hunk 2 above)
Comment (tustvold) [reply parquet/src/arrow/array_reader/map_array.rs:50]: I'd say it is more, if this level is nullable the schema is inconsistent as there is no way to represent that in parquet :sweat_smile: The schema inference logic will never generate this - https://github.com/apache/arrow-rs/blob/master/parquet/src/arrow/schema/complex.rs#L350

File: parquet/src/arrow/array_reader/map_array.rs [hunk 4]
```diff
@@ -43,15 +35,43 @@ impl MapArrayReader {
...
+        let struct_rep_level = rep_level + 1;
+
+        let element = match &data_type {
+            ArrowType::Map(element, _) => match element.data_type() {
+                ArrowType::Struct(fields) if fields.len() == 2 => {
+                    // The inner map field must always non-nullable (#1697)
+                    assert!(!element.is_nullable(), "map struct cannot be nullable");
+                    element
+                }
+                _ => unreachable!("expected struct with two fields"),
```
[the remark this answers is not in the document]
Comment (tustvold) [reply parquet/src/arrow/array_reader/map_array.rs:53]: This isn't a public API so I could change it, but that also makes me less inclined to do so. It is a bug in the builder if you run into this

File: parquet/src/arrow/array_reader/map_array.rs [hunk 5]
```diff
@@ -65,131 +85,122 @@ impl ArrayReader for MapArrayReader {
...
-        let array_data = unsafe { array_data.build_unchecked() };
-
-        Ok(Arc::new(MapArray::from(array_data)))
+        // A MapArray is just a ListArray with a StructArray child
+        // we can therefore just alter the ArrayData
+        let array = self.reader.consume_batch().unwrap();
+        let data = array.data().clone();
+        let builder = data.into_builder().data_type(self.data_type.clone());
+        Ok(Arc::new(MapArray::from(unsafe {
+            builder.build_unchecked()
```
[the remark this answers is not in the document]
Comment (tustvold) [reply parquet/src/arrow/array_reader/map_array.rs:98]: Pretty much, will add a doc comment

File: parquet/src/arrow/array_reader/map_array.rs (hunk 2 above)
Comment (alamb) [reply parquet/src/arrow/array_reader/map_array.rs:50]: maybe worth a comment there

File: parquet/src/arrow/array_reader/map_array.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (alamb) [reply parquet/src/arrow/array_reader/map_array.rs:53]: makes sense 👍

File: parquet/src/arrow/array_reader/map_array.rs (hunk 3 above)
Comment (alamb) [reply parquet/src/arrow/array_reader/map_array.rs:196]: Yeah, sorry I probably should have phrased that as "I suggest this test also validate the first two entries are null"

File: parquet/src/arrow/array_reader/map_array.rs [hunk 6]
```diff
@@ -193,6 +201,8 @@ mod tests {
         for maybe_record_batch in record_batch_reader {
             let record_batch = maybe_record_batch.expect("Getting current batch");
             let col = record_batch.column(0);
+            assert!(col.is_null(0));
```
Comment (alamb) [inline parquet/src/arrow/array_reader/map_array.rs:204]: ❤️

**Comment by d-willis** [comment]
Thanks for the quick turn around on this!

**Pull request merged**